src/main/scala/org/graphframes/GraphFrame.scala

```diff
@@ -476,8 +475,15 @@ class GraphFrame private(
       indexedVertices.select(
         col(ATTR + "." + ID).cast("long").as(LONG_ID), col(ATTR + "." + ID).as(ID), col(ATTR))
     } else {
-      val indexedVertices = zipWithUniqueId(vertices)
-      indexedVertices.select(col("uniq_id").as(LONG_ID), col("row." + ID).as(ID), col("row").as(ATTR))
+      val withLongIds = vertices.select(ID)
+        .repartition(col(ID))
+        .distinct()
+        .sortWithinPartitions(ID)
+        .withColumn(LONG_ID, monotonically_increasing_id())
+        .persist(StorageLevel.MEMORY_AND_DISK)
+      vertices.select(col(ID), nestAsCol(vertices, ATTR))
+        .join(withLongIds, ID)
+        .select(LONG_ID, ID, ATTR)
```

Comment:
I wonder if it is worth the extra effort to optimize for the newer Spark release.
if `monotonically_increasing_id` not unique is only in earlier ones, perhaps we should create a wrapper for it instead of penalizing all versions with extra repartition?